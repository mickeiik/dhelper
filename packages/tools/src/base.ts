import { z } from 'zod';
import { ResultSchema } from '@app/schemas';

export abstract class Tool<
    TInput extends z.ZodType,
    TOutput extends z.ZodType
> {
    abstract id: string;
    abstract name: string;
    abstract inputSchema: TInput;
    abstract outputSchema: TOutput;

    // Optional properties that tools commonly implement
    abstract description: string;
    category?: string;
    examples?: unknown;

    // Optional initialization method
    initialize?(context: any): Promise<void>;

    // Create the result schema for this tool
    get resultSchema() {
        return ResultSchema(this.outputSchema);
    }

    async execute(rawInput: z.input<TInput>): Promise<z.infer<typeof this.resultSchema>> {
        try {
            // Validate input
            const input = this.inputSchema.parse(rawInput);

            // Execute with validated input
            const result = await this.executeValidated(input);

            // Validate and return the result
            return this.resultSchema.parse(result);
        } catch (error) {
            // Return error result
            return this.resultSchema.parse({
                success: false,
                error: {
                    message: error instanceof Error ? error.message : String(error),
                    code: 'TOOL_EXECUTION_ERROR',
                    details: { originalError: error }
                }
            });
        }
    }

    protected abstract executeValidated(
        input: z.output<TInput>
    ): Promise<z.infer<typeof this.resultSchema>>;
}