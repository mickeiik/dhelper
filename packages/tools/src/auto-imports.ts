// Auto-generated file - DO NOT EDIT MANUALLY
// This file is generated by scripts/generate-tool-imports.js
// It ensures all tool module augmentations are loaded for TypeScript autocomplete

import '@tools/hello-world';
import '@tools/ocr';
import '@tools/screenshot';

// This export is required to make this a module
export {};
