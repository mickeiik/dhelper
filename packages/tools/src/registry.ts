// Manual tool type registry - update when adding new tools
export interface ToolRegistry {
  'hello-world': {
    input: import('@tools/hello-world').HelloWorldToolInput;
    output: import('@tools/hello-world').HelloWorldToolOutput;
  };
  'ocr': {
    input: import('@tools/ocr').TesseractOcrToolInput;
    output: import('@tools/ocr').TesseractOcrToolOutput;
  };
  'screenshot': {
    input: import('@tools/screenshot').ScreenshotToolInput;
    output: import('@tools/screenshot').ScreenshotToolOutput;
  };
  'screen-region-selector': {
    input: import('@tools/screen-region-selector').ScreenRegionSelectorInput;
    output: import('@tools/screen-region-selector').ScreenRegionSelectorOutput;
  };
  'template-matcher': {
    input: import('@tools/template-matcher').TemplateMatcherInput;
    output: import('@tools/template-matcher').TemplateMatcherOutput;
  };
}

export type ToolId = keyof ToolRegistry;

// Input/Output types for registered tools
export type ToolInput<T extends ToolId> = ToolRegistry[T]['input'];
export type ToolOutput<T extends ToolId> = ToolRegistry[T]['output'];