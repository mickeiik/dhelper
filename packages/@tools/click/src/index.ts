import type { OverlayService } from '@app/overlay';
import { OVERLAY_STYLES } from '@app/overlay';
import { OverlayShapeSchema, ClickInputSchema, ClickOutputSchema, ToolResult, PointSchema } from '@app/schemas';
import { mouse, Button, sleep } from '@nut-tree-fork/nut-js';
import { Tool } from '@app/tools';
import { z } from 'zod';

// Type aliases for convenience
type ClickToolInput = z.infer<typeof ClickInputSchema>;
type ClickToolOutput = z.infer<typeof ClickOutputSchema>;
type ClickResult = ToolResult<typeof ClickOutputSchema>;
type OverlayShape = z.infer<typeof OverlayShapeSchema>;
export class ClickTool extends Tool<typeof ClickInputSchema, typeof ClickOutputSchema> {
  id = 'click' as const;
  name = 'Click Tool';
  description = 'Click at screen positions with multiple click methods and input types';
  category = 'Input';

  inputSchema = ClickInputSchema;
  outputSchema = ClickOutputSchema;

  private overlayService?: OverlayService;

  examples = [
    {
      name: 'Simple Click',
      description: 'Click at specific coordinates',
      inputs: {
        x: 500,
        y: 300,
        showVisualIndicator: true
      }
    },
    {
      name: 'Click in Region Center',
      description: 'Click in the center of a region',
      inputs: {
        region: { x: 100, y: 100, width: 200, height: 100 },
      }
    },
    {
      name: 'Double Click',
      description: 'Perform a double click',
      inputs: {
        x: 400,
        y: 250,
        clicks: 2,
        delay: 50
      }
    },
    {
      name: 'Right Click',
      description: 'Perform a right click',
      inputs: {
        x: 600,
        y: 400,
        button: 'right'
      }
    },
    {
      name: 'Click Template Match Result',
      description: 'Click on a template match result from previous step',
      inputs: {
        region: {
          x: { $ref: '{{previous:template-matcher.location.x}}' },
          y: { $ref: '{{previous:template-matcher.location.y}}' },
          width: { $ref: '{{previous:template-matcher.location.width}}' },
          height: { $ref: '{{previous:template-matcher.location.height}}' }
        }
      }
    },
    {
      name: 'Click Selected Region',
      description: 'Click in the center of a region selected with screen-region-selector',
      inputs: {
        region: {
          x: { $ref: '{{previous:screen-region-selector.x}}' },
          y: { $ref: '{{previous:screen-region-selector.y}}' },
          width: { $ref: '{{previous:screen-region-selector.width}}' },
          height: { $ref: '{{previous:screen-region-selector.height}}' }
        }
      }
    }
  ];

  async initialize(context: any) {
    this.overlayService = context.overlayService;
    return;
  }

  async executeValidated(input: ClickToolInput): Promise<ClickResult> {
    // Determine click position
    const position = this.calculateClickPosition(input);

    // Get click configuration
    const button = this.getMouseButton(input.button);

    // Show visual indicator before clicking
    if (input.showVisualIndicator !== false) {
      await this.showVisualIndicator(position, input.indicatorTimeout);
    }

    // Perform the click using selected method
    await this.clickDefault(position, button, input.clicks, input.delay);

    return {
      success: true,
      data: position // Return the clicked position {x, y}
    };
  }

  private calculateClickPosition(input: ClickToolInput): ClickToolOutput {
    const isRectangle = 'width' in input;

    if (isRectangle) {
      return {
        x: input.x + Math.floor(input.width / 2),
        y: input.y + Math.floor(input.height / 2)
      };
    }

    return { x: input.x, y: input.y };
  }

  private getMouseButton(button: string): Button {
    switch (button.toLowerCase()) {
      case 'right': return Button.RIGHT;
      case 'middle': return Button.MIDDLE;
      case 'left':
      default: return Button.LEFT;
    }
  }

  private async clickDefault(
    position: z.infer<typeof PointSchema>,
    button: Button,
    clicks: number,
    delay: number
  ): Promise<void> {
    // Move to position and click
    await mouse.setPosition(position);
    await sleep(50);

    for (let i = 0; i < clicks; i++) {
      await mouse.pressButton(button);
      await mouse.releaseButton(button);
      if (i < clicks - 1) {
        await sleep(delay);
      }
    }
  }

  private async showVisualIndicator(
    position: z.infer<typeof PointSchema>,
    timeout: number
  ): Promise<void> {
    if (!this.overlayService) {
      console.warn('[Click Tool] Overlay service not available, skipping visual indicator');
      return;
    }

    try {
      // Create overlay window
      const overlay = await this.overlayService.createOverlay({
        showInstructions: false,
        timeout: timeout,
        clickThrough: true
      });

      // Create crosshair and circle at click position
      const indicatorSize = 20;
      const shapes: OverlayShape[] = [
        // Circle
        {
          id: 'click-circle',
          type: 'rectangle',
          bounds: {
            x: position.x - indicatorSize / 2,
            y: position.y - indicatorSize / 2,
            width: indicatorSize,
            height: indicatorSize
          },
          style: {
            ...OVERLAY_STYLES.HIGHLIGHT,
            color: '#ff0000',
            lineWidth: 3,
            fillColor: 'rgba(255, 0, 0, 0.2)'
          },
          label: 'Click Target',
          labelPosition: 'top'
        }
      ];

      // Crosshair lines
      shapes.push(
        // Horizontal line
        {
          id: 'click-line-h',
          type: 'rectangle',
          bounds: {
            x: position.x - 10,
            y: position.y - 1,
            width: 20,
            height: 2
          },
          style: {
            color: '#ff0000',
            lineWidth: 2,
            fillColor: '#ff0000'
          }
        },
        // Vertical line
        {
          id: 'click-line-v',
          type: 'rectangle',
          bounds: {
            x: position.x - 1,
            y: position.y - 10,
            width: 2,
            height: 20
          },
          style: {
            color: '#ff0000',
            lineWidth: 2,
            fillColor: '#ff0000'
          }
        }
      );

      // Draw shapes
      await overlay.drawShapes(shapes);

      // Show the overlay
      await overlay.show();

    } catch (error) {
      console.error('[Click Tool] Failed to show visual indicator:', error);
    }
  }
}