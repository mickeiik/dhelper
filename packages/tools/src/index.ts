// packages/tools/src/index.ts
export * from './base.js';

import type { OverlayService } from '@app/types';
import { ToolExecutionError, ErrorLogger } from '@app/types';
import { ToolMetadataSchema, ToolResult } from '@app/schemas';
import { z } from 'zod';
import { Tool } from './base.js';

/**
 * Tool imports - Add new tools here
 * 
 * To add a new tool:
 * 1. Import your tool class here
 * 2. Add it to the AVAILABLE_TOOLS array below
 */
import { HelloWorldTool } from '@tools/hello-world';
import { TesseractOcrTool } from '@tools/ocr';
import { ScreenRegionSelectorTool } from '@tools/screen-region-selector';
import { ScreenshotTool } from '@tools/screenshot';
import { TemplateMatcherTool } from '@tools/template-matcher';
import { ClickTool } from '@tools/click';

interface ToolRegistration {
    tool?: Tool<z.ZodType, z.ZodType>;
    loader?: () => Promise<Tool<z.ZodType, z.ZodType>>;
    initialized: boolean;
    loading: boolean;
}

/**
 * Registry of all available tools
 * 
 * Add your imported tool class to this array to make it available in the system.
 * Tools are automatically instantiated and registered when the ToolManager initializes.
 */
const AVAILABLE_TOOLS = [
    HelloWorldTool,
    TesseractOcrTool,
    ScreenRegionSelectorTool,
    ScreenshotTool,
    TemplateMatcherTool,
    ClickTool
] as const;

// Auto-generated types from AVAILABLE_TOOLS
type ToolInstances = InstanceType<typeof AVAILABLE_TOOLS[number]>;

// Auto-generated tool registry types
export type ToolId = ToolInstances['id'];

export type ToolInput<T extends ToolId> = Extract<ToolInstances, { id: T }> extends { 
    inputSchema: infer S extends z.ZodType 
} ? z.infer<S> : never;

export type ToolOutput<T extends ToolId> = Extract<ToolInstances, { id: T }> extends { 
    outputSchema: infer S extends z.ZodType 
} ? z.infer<S> : never;

export class ToolManager {
    private tools = new Map<string, ToolRegistration>();
    private overlayService?: OverlayService;
    private templateManager?: import('@app/types').TemplateManager;
    private logger = new ErrorLogger('ToolManager');
    private loadingPromises = new Map<string, Promise<Tool<z.ZodType, z.ZodType>>>();

    async autoDiscoverTools() {
        // Register all available tools
        for (const ToolClass of AVAILABLE_TOOLS) {
            try {
                const tool = new ToolClass();
                
                // Validate tool structure at registration time
                const metadata = {
                    id: tool.id,
                    name: tool.name,
                    description: tool.description,
                    category: tool.category,
                    inputSchema: tool.inputSchema,
                    outputSchema: tool.outputSchema,
                    resultSchema: tool.resultSchema,
                    examples: tool.examples
                };
                
                ToolMetadataSchema.parse(metadata);
                
                this.tools.set(tool.id, {
                    tool,
                    initialized: false,
                    loading: false
                });
            } catch (error) {
                const errorMessage = error instanceof z.ZodError 
                    ? `Tool ${ToolClass.name} failed validation: ${error.message}`
                    : `Failed to instantiate tool ${ToolClass.name}`;
                    
                this.logger.logError(new ToolExecutionError(
                    errorMessage,
                    ToolClass.name,
                    { originalError: error }
                ));
            }
        }
    }

    /**
     * Register a tool with async loading
     * Useful for tools with heavy dependencies that should be loaded on-demand
     */
    registerAsyncTool(id: string, loader: () => Promise<Tool<z.ZodType, z.ZodType>>) {
        this.tools.set(id, {
            loader,
            initialized: false,
            loading: false
        });
    }

    /**
     * Load a tool if it's not already loaded
     */
    private async ensureToolLoaded(id: string): Promise<Tool<z.ZodType, z.ZodType>> {
        const registration = this.tools.get(id);
        if (!registration) {
            throw new ToolExecutionError(`Tool with id "${id}" not found`, id);
        }

        // If tool is already loaded, return it
        if (registration.tool) {
            return registration.tool;
        }

        // If tool is currently loading, wait for it
        if (registration.loading && this.loadingPromises.has(id)) {
            return this.loadingPromises.get(id)!;
        }

        // If tool has a loader, load it
        if (registration.loader) {
            registration.loading = true;
            
            const loadingPromise = (async () => {
                try {
                    const tool = await registration.loader!();
                    
                    // Validate the loaded tool
                    const metadata = {
                        id: tool.id,
                        name: tool.name,
                        description: tool.description,
                        category: tool.category,
                        inputSchema: tool.inputSchema,
                        outputSchema: tool.outputSchema,
                        resultSchema: tool.resultSchema,
                        examples: tool.examples
                    };
                    
                    ToolMetadataSchema.parse(metadata);
                    
                    registration.tool = tool;
                    registration.loading = false;
                    this.loadingPromises.delete(id);
                    
                    return tool;
                } catch (error) {
                    registration.loading = false;
                    this.loadingPromises.delete(id);
                    
                    const errorMessage = error instanceof z.ZodError 
                        ? `Tool ${id} failed validation after loading: ${error.message}`
                        : `Failed to load tool ${id}`;
                        
                    const toolError = new ToolExecutionError(errorMessage, id, { originalError: error });
                    this.logger.logError(toolError);
                    throw toolError;
                }
            })();
            
            this.loadingPromises.set(id, loadingPromise);
            return loadingPromise;
        }

        throw new ToolExecutionError(`Tool "${id}" has no instance or loader`, id);
    }

    async runTool<T extends ToolId>(
        id: T, 
        inputs: ToolInput<T>
    ): Promise<ToolResult<z.ZodType>> {
        try {
            // Ensure tool is loaded (handles both sync and async tools)
            const tool = await this.ensureToolLoaded(id);
            
            // Initialize tool if needed
            const registration = this.tools.get(id)!;
            if (!registration.initialized && tool.initialize) {
                await this.initializeTool(id);
            }

            // Tool.execute already handles Zod validation and error wrapping
            return await tool.execute(inputs);
        } catch (error) {
            const toolError = error instanceof ToolExecutionError ? error : new ToolExecutionError(`Tool "${id}" execution failed`, id, { originalError: error, inputs });
            this.logger.logError(toolError);
            throw toolError;
        }
    }

    setOverlayService(overlayService: OverlayService) {
        this.overlayService = overlayService;
    }

    setTemplateManager(templateManager: import('@app/types').TemplateManager) {
        this.templateManager = templateManager;
    }

    private async initializeTool(id: string) {
        const registration = this.tools.get(id);
        if (!registration || registration.initialized) return;

        // Ensure tool is loaded before initializing
        const tool = await this.ensureToolLoaded(id);
        if (!tool.initialize) return;

        try {
            const initContext = {
                overlayService: this.overlayService,
                templateManager: this.templateManager
            };
            await tool.initialize(initContext);
            registration.initialized = true;
        } catch (error) {
            const toolError = new ToolExecutionError(`Failed to initialize tool "${id}"`, id, { originalError: error });
            this.logger.logError(toolError);
            throw toolError;
        }
    }

    async getTools(): Promise<Tool<z.ZodType, z.ZodType>[]> {
        const tools: Tool<z.ZodType, z.ZodType>[] = [];
        
        for (const [id] of this.tools.entries()) {
            try {
                const tool = await this.ensureToolLoaded(id);
                tools.push(tool);
            } catch (error) {
                this.logger.logError(new ToolExecutionError(
                    `Failed to load tool ${id}`,
                    id,
                    { originalError: error }
                ));
            }
        }
        
        return tools;
    }

    async getToolsMetadata(): Promise<z.infer<typeof ToolMetadataSchema>[]> {
        const metadata: z.infer<typeof ToolMetadataSchema>[] = [];
        
        for (const [id] of this.tools.entries()) {
            try {
                const tool = await this.ensureToolLoaded(id);
                metadata.push(ToolMetadataSchema.parse({
                    id: tool.id,
                    name: tool.name,
                    description: tool.description,
                    category: tool.category,
                    inputSchema: tool.inputSchema,
                    outputSchema: tool.outputSchema,
                    resultSchema: tool.resultSchema,
                    examples: tool.examples
                }));
            } catch (error) {
                this.logger.logError(new ToolExecutionError(
                    `Failed to load metadata for tool ${id}`,
                    id,
                    { originalError: error }
                ));
            }
        }
        
        return metadata;
    }
}

// Re-export shared types for convenience
export type { ToolMetadata } from '@app/types';