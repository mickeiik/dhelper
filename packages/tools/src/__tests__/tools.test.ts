import { describe, test, expect, beforeEach, vi } from 'vitest';
import { z } from 'zod';
import { Tool, ToolId, ToolManager } from '../src/index.js';
import { ResultSchema, ToolMetadataSchema } from '../../schemas/src';
import { afterEach } from 'node:test';

const MockInputSchema = z.object({
    message: z.string(),
    value: z.number().optional()
});

const MockOutputSchema = z.object({
    result: z.string(),
    processed: z.boolean()
});

class MockTool extends Tool<typeof MockInputSchema, typeof MockOutputSchema> {
    id = 'mock-tool' as const;
    name = 'Mock Tool';
    description = 'A mock tool for testing';
    category = 'Test';

    inputSchema = MockInputSchema;
    outputSchema = MockOutputSchema;

    async executeValidated(input: z.infer<typeof MockInputSchema>) {
        return {
            success: true as const,
            data: {
                result: `Processed: ${input.message}`,
                processed: true
            }
        };
    }
}

const initializeMock = vi.fn()

class MockToolWithInit extends Tool<typeof MockInputSchema, typeof MockOutputSchema> {
    id = 'mock-tool-init' as const;
    name = 'Mock Tool with Init';
    description = 'dfsf';
    inputSchema = MockInputSchema;
    outputSchema = MockOutputSchema;

    async initialize(context: any): Promise<void> {
        initializeMock()
    }

    async executeValidated(input: z.infer<typeof MockInputSchema>) {
        return {
            success: true as const,
            data: {
                result: `Processed: ${input.message}`,
                processed: true
            }
        };
    }
}

class MockFailingTool extends Tool<typeof MockInputSchema, typeof MockOutputSchema> {
    id = 'mock-failing-tool' as const;
    name = 'Mock Failing Tool';
    description = 'A tool that always fails';
    category = 'Test';

    inputSchema = MockInputSchema;
    outputSchema = MockOutputSchema;

    async executeValidated(): Promise<z.infer<typeof this.resultSchema>> {
        throw new Error('Tool execution failed');
    }
}

describe('Tool Base Class', () => {
    let tool: MockTool;

    beforeEach(() => {
        tool = new MockTool();
    });

    describe('Tool Properties', () => {
        test('should have required properties', () => {
            expect(tool.id).toBe('mock-tool');
            expect(tool.name).toBe('Mock Tool');
            expect(tool.description).toBe('A mock tool for testing');
            expect(tool.category).toBe('Test');
            expect(tool.inputSchema).toBe(MockInputSchema);
            expect(tool.outputSchema).toBe(MockOutputSchema);
        });

        test('should generate correct result schema', () => {
            const resultSchema = tool.resultSchema;
            expect(resultSchema).toBeDefined();

            const result = {
                success: true,
                data: {
                    result: 'result string',
                    processed: true
                }
            };

            const parseResult = () => tool.resultSchema.parse(result);

            expect(parseResult).not.toThrow();
        });
    });

    describe('Tool Execution', () => {
        test('should execute with valid input', async () => {
            const input = { message: 'test message' };
            const result = await tool.execute(input);

            expect(result.success).toBe(true);
            if (result.success) {
                expect(result.data.result).toBe('Processed: test message');
                expect(result.data.processed).toBe(true);
            }
        });

        test('should validate input schema', async () => {
            const invalidInput = { wrongField: 'test' };
            const result = await tool.execute(invalidInput);

            expect(result.success).toBe(false);
            if (!result.success) {
                expect(result.error.message).toContain('message');
                expect(result.error.code).toBe('TOOL_EXECUTION_ERROR');
            }
            expect(result).toMatchSnapshot();
        });

        test('should handle optional input fields', async () => {
            const input = { message: 'test', value: 42 };
            const result = await tool.execute(input);

            expect(result.success).toBe(true);
            if (result.success) {
                expect(result.data.result).toBe(`Processed: ${input.message}`);
            }
        });
    });
});

describe('ToolManager', () => {
    let toolManager: ToolManager;

    beforeEach(() => {
        toolManager = new ToolManager();
    });

    afterEach(() => {
        vi.restoreAllMocks();
    });

    describe('Tool Discovery and Registration', () => {
        beforeEach(() => {
            vi.restoreAllMocks();
        });

        test('should auto-discover tools from loaders', async () => {
            const mockToolLoader = () => Promise.resolve(MockTool);
            await toolManager.autoDiscoverTools([mockToolLoader]);

            const tools = await toolManager.getTools();
            expect(tools).toHaveLength(1);
            expect(tools[0].id).toBe('mock-tool');
        });

        test('should validate tool metadata during discovery', async () => {
            const consoleSpy = vi.spyOn(console, 'error').mockImplementation(() => { });
            const { mockParse } = vi.hoisted(() => {
                return {
                    //Mock ToolMetadataSchema.parse but keep original behavior (should not add an invalid tool to the manager)
                    mockParse: vi.fn((input) => ToolMetadataSchema.parse(input))
                }
            });

            vi.mock('@app/schemas', async (importOriginal) => {
                const mod = await importOriginal<typeof import('@app/schemas')>()
                return {
                    ResultSchema: mod.ResultSchema, //Needed to validate tool 'resultSchema'
                    ToolMetadataSchema: {
                        parse: mockParse
                    }
                }
            });

            const emptyObjectSchema = z.object({});
            // Create an invalid tool class
            class InvalidTool extends Tool<typeof emptyObjectSchema, typeof emptyObjectSchema> {
                id = ''; // Invalid empty id
                name = 'Invalid Tool';
                description = 'Invalid tool';
                inputSchema = emptyObjectSchema;
                outputSchema = emptyObjectSchema;

                async executeValidated(): Promise<z.infer<typeof this.resultSchema>> {
                    return { success: true, data: {} };
                }
            }

            const invalidToolLoader = () => Promise.resolve(InvalidTool);
            await toolManager.autoDiscoverTools([invalidToolLoader]);
            const tools = await toolManager.getTools();

            //Should try to parse the invalid tool with zod
            expect(mockParse).toHaveBeenCalledTimes(1)
            expect(JSON.stringify(mockParse.mock.calls[0][0])).toBe(
                JSON.stringify({
                    id: '',
                    name: 'Invalid Tool',
                    description: 'Invalid tool',
                    inputSchema: emptyObjectSchema,
                    outputSchema: emptyObjectSchema,
                    resultSchema: ResultSchema(emptyObjectSchema),
                })
            );

            //Should not add an invalid tool to the manager
            expect(tools).toHaveLength(0);
            expect(consoleSpy).toHaveBeenCalledWith(
                expect.stringContaining('[ToolManager] Tool failed validation:'),
                expect.any(Error)
            );

            mockParse.mockRestore()
        });

        test('should handle tool loading errors', async () => {
            const consoleSpy = vi.spyOn(console, 'error').mockImplementation(() => { });

            const failingLoader = () => Promise.reject(new Error('Load failed'));
            await toolManager.autoDiscoverTools([failingLoader]);

            const tools = await toolManager.getTools();
            expect(tools).toHaveLength(0);
            expect(consoleSpy).toHaveBeenCalled();

            consoleSpy.mockRestore();
        });
    });

    describe('Async Tool Loading', () => {
        test('should register async tools', async () => {
            const toolManagerToolsMap = toolManager['tools'];
            const asyncLoader = () => Promise.resolve(new MockTool());

            expect(toolManagerToolsMap).toHaveLength(0)

            toolManager.registerAsyncTool('async-tool', asyncLoader);

            // Tool should be registered but not yet loaded
            const registeredAsyncTool = toolManager['tools'].get('async-tool');

            expect(toolManagerToolsMap).toHaveLength(1)
            expect(registeredAsyncTool?.loader).toBe(asyncLoader)
            expect(registeredAsyncTool?.initialized).toBe(false)
            expect(registeredAsyncTool?.loading).toBe(false)
        });

        test('should load async tools on demand', async () => {
            const toolManagerToolsMap = toolManager['tools'];
            const asyncLoader = () => Promise.resolve(new MockTool());

            expect(toolManagerToolsMap).toHaveLength(0);

            toolManager.registerAsyncTool('mock-tool', asyncLoader);

            const registeredAsyncTool = toolManager['tools'].get('mock-tool');
            expect(registeredAsyncTool?.tool).toBeUndefined() //Tool not loaded yet


            const input = { message: 'test async' };
            await toolManager.runTool('mock-tool' as ToolId, input); //Should load the tool

            expect(registeredAsyncTool?.tool).toBeDefined() //Tool loaded
        });

        test('should handle concurrent async tool loading', async () => {
            const asyncLoader = async () => {
                await new Promise(resolve => setTimeout(resolve, 10));
                return new MockTool();
            };

            toolManager.registerAsyncTool('mock-tool', asyncLoader);

            // Start multiple concurrent operations
            const results = await Promise.all([
                toolManager.runTool('mock-tool' as ToolId, { message: 'test1' }),
                toolManager.runTool('mock-tool' as ToolId, { message: 'test2' }),
                toolManager.runTool('mock-tool' as ToolId, { message: 'test3' })
            ]);

            const registeredAsyncTool = toolManager['tools'];

            // Tool should only be loaded once
            expect(registeredAsyncTool).toHaveLength(1);
            expect(results).toHaveLength(3);
            results.forEach(result => {
                expect(result.success).toBe(true);
            });
        });

        test('should throw if an async registered tool does not have a loader', async () => {
            const consoleSpy = vi.spyOn(console, 'error').mockImplementation(() => { });

            const asyncLoader = () => Promise.resolve(new MockTool());
            // await toolManager.autoDiscoverTools([mockToolLoader]);
            toolManager.registerAsyncTool('mock-tool', asyncLoader);


            toolManager['tools'].get('mock-tool')!.loader = undefined

            await toolManager.getTools();

            expect(consoleSpy).toHaveBeenCalledTimes(1);
        })

        test('should throw if a async loaded tool hase a wrong ToolMetadataSche', async () => {
            const consoleSpy = vi.spyOn(console, 'error').mockImplementation(() => {});

            const emptyObjectSchema = z.object({});
            class InvalidTool extends Tool<typeof emptyObjectSchema, typeof emptyObjectSchema> {
                id = 'mock-tool'; // Invalid empty id
                name = 12 as unknown as string;
                description = 'Invalid tool';
                inputSchema = emptyObjectSchema;
                outputSchema = emptyObjectSchema;

                async executeValidated(): Promise<z.infer<typeof this.resultSchema>> {
                    return { success: true, data: {} };
                }
            }
            const asyncLoader = () => Promise.resolve(new InvalidTool());

            toolManager.registerAsyncTool('mock-tool', asyncLoader);

            await toolManager.getTools();

            expect(consoleSpy).toHaveBeenCalledTimes(2);
            expect(consoleSpy).toMatchSnapshot();
        })
    });

    describe('Tool Initialization', () => {
        test('should initialize tools that require it', async () => {
            const mockToolLoader = () => Promise.resolve(MockToolWithInit);
            await toolManager.autoDiscoverTools([mockToolLoader]);

            const input = { message: 'test init' };
            const result = await toolManager.runTool('mock-tool-init' as ToolId, input);

            expect(initializeMock).toHaveBeenCalledTimes(1)
            expect(result.success).toBe(true);
        });

        test('should handle initialization errors', async () => {
            const consoleSpy = vi.spyOn(console, 'error').mockImplementation(() => { });

            class FailingInitTool extends MockToolWithInit {
                async initialize() {
                    throw new Error('Initialization failed');
                }
            }

            const mockToolLoader = () => Promise.resolve(FailingInitTool);
            await toolManager.autoDiscoverTools([mockToolLoader]);

            const input = { message: 'test' };
            await expect(toolManager.runTool('mock-tool-init' as ToolId, input))
                .rejects.toThrow('Failed to initialize tool');
            expect(consoleSpy).toHaveBeenCalledTimes(2);
        });
    });

    describe('Tool Execution via Manager', () => {
        beforeEach(async () => {
            const mockToolLoader = () => Promise.resolve(MockTool);
            await toolManager.autoDiscoverTools([mockToolLoader]);
        });

        test('should run tools successfully', async () => {
            const input = { message: 'manager test' };
            const result = await toolManager.runTool('mock-tool' as ToolId, input);

            expect(result.success).toBe(true);
            if (result.success) {
                const resultData = result.data as z.infer<typeof MockOutputSchema>;
                expect(resultData.result).toBe('Processed: manager test');
            }
        });

        test('should throw error for non-existent tools', async () => {
            const input = { message: 'test' };
            await expect(toolManager.runTool('non-existent' as any, input))
                .rejects.toThrow('Tool with id "non-existent" not found');
        });

        test('should handle tool execution failures', async () => {
            const failingToolLoader = () => Promise.resolve(MockFailingTool);
            await toolManager.autoDiscoverTools([failingToolLoader]);

            const input = { message: 'test' };
            const result = await toolManager.runTool('mock-failing-tool' as ToolId, input);

            expect(result.success).toBe(false);
            if (!result.success) {
                expect(result.error.message).toBe('Tool execution failed');
            }
        });
    });

    describe('Tool Metadata', () => {
        beforeEach(async () => {
            const mockToolLoader = () => Promise.resolve(MockTool);
            await toolManager.autoDiscoverTools([mockToolLoader]);
        });

        test('should return tools metadata', async () => {
            const metadata = await toolManager.getToolsMetadata();

            expect(metadata).toHaveLength(1);
            expect(metadata[0]).toMatchObject({
                id: 'mock-tool',
                name: 'Mock Tool',
                description: 'A mock tool for testing',
                category: 'Test'
            });

            // Validate against schema
            expect(() => ToolMetadataSchema.parse(metadata[0])).not.toThrow();
        });

        test('should handle metadata loading errors gracefully', async () => {
            const consoleSpy = vi.spyOn(console, 'error').mockImplementation(() => { });

            // Register a tool that will fail to load
            const failingLoader = () => Promise.reject(new Error('Load failed'));
            toolManager.registerAsyncTool('failing-tool', failingLoader);

            const metadata = await toolManager.getToolsMetadata();

            // Should return metadata for working tools only
            expect(metadata).toHaveLength(1);
            expect(metadata[0].id).toBe('mock-tool');
            expect(consoleSpy).toHaveBeenCalled();
        });
    });

    describe('Service Integration', () => {
        test('should set overlay service', () => {
            const mockOverlayService = {} as any;
            expect(() => toolManager.setOverlayService(mockOverlayService)).not.toThrow();
        });

        test('should set template manager', () => {
            const mockTemplateManager = {} as any;
            expect(() => toolManager.setTemplateManager(mockTemplateManager)).not.toThrow();
        });
    });
});
