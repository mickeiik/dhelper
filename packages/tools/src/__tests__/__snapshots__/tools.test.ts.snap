// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tool Base Class > Tool Execution > should validate input schema 1`] = `
{
  "error": {
    "code": "TOOL_EXECUTION_ERROR",
    "details": {
      "originalError": [ZodError: [
  {
    "expected": "string",
    "code": "invalid_type",
    "path": [
      "message"
    ],
    "message": "Invalid input: expected string, received undefined"
  }
]],
    },
    "message": "[
  {
    "expected": "string",
    "code": "invalid_type",
    "path": [
      "message"
    ],
    "message": "Invalid input: expected string, received undefined"
  }
]",
  },
  "success": false,
}
`;

exports[`ToolManager > Async Tool Loading > should throw if a async loaded tool hase a wrong ToolMetadataSche 1`] = `
[MockFunction error] {
  "calls": [
    [
      "[ToolManager] Tool mock-tool failed validation after loading: [
  {
    "expected": "string",
    "code": "invalid_type",
    "path": [
      "name"
    ],
    "message": "Invalid input: expected string, received number"
  }
]",
      [Error: Tool mock-tool failed validation after loading: [
  {
    "expected": "string",
    "code": "invalid_type",
    "path": [
      "name"
    ],
    "message": "Invalid input: expected string, received number"
  }
]],
    ],
    [
      "[ToolManager] Failed to load tool mock-tool",
      [Error: Tool mock-tool failed validation after loading: [
  {
    "expected": "string",
    "code": "invalid_type",
    "path": [
      "name"
    ],
    "message": "Invalid input: expected string, received number"
  }
]],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
